testing/src/main/java/io/grpc/testing/GrpcServerRule.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing;
+
+import com.google.common.base.Supplier;
+
+import io.grpc.BindableService;
+import io.grpc.ManagedChannel;
+import io.grpc.Server;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.AbstractStub;
+
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TestRule;
+
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@code GrpcServerRule} is a JUnit {@link TestRule} that starts an in-process gRPC service with
+ * the given service implementation. It is particularly useful for mocking out external gRPC-based
+ * services and asserting that the expected requests were made.
+ *
+ * <p>An {@link AbstractStub} can be created against this service by using the
+ * {@link ManagedChannel} provided by {@link GrpcServerRule#channel()}.
+ *
+ * @param <T> the type of the service impl
+ */
+public class GrpcServerRule<T extends BindableService> extends ExternalResource {
+
+  private final Supplier<T> serviceImplSupplier;
+
+  private ManagedChannel channel;
+  private Server server;
+  private String serverName;
+  private T serviceImpl;
+
+  public GrpcServerRule(Supplier<T> serviceImplSupplier) {
```

Comment:
I think it may be nicer to use `io.grpc.util.MutableHandlerRegistry` in the server and allow someone to create and add the service later. That's what the [RouteGuideClientTest does](https://github.com/grpc/grpc-java/blob/master/examples/src/test/java/io/grpc/examples/routeguide/RouteGuideClientTest.java), for instance. If the test class uses the same service throughout, then it can just use a `@Before` to add it.